### PR TITLE
[DS][46/n] Create NewlyUpdatedCondition

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

As title. When testing this, I realized that two latent bugs were impacting the correctness here.

Essentially, when computing the set of things changed since a given cursor inside the CachingInstanceQueryer, we return a new cursor representing the latest storage id across all records that were used to make the decision (i.e. the target asset's max storage id and all of its parents max storage ids).

However, we were shortcutting this logic in the case that a child had no parent assets, meaning an asset with no parents would never have its cursor advance!

After fixing that, I noticed that a separate test started failing, dealing with how we calculate which asset partitions are missing when you update the partitions definition.

The secondary bug (only present in the legacy missing rule, not the new version), happened because the legacy rule stores its knowledge of what's NOT missing inside the cursor, rather than just using the asset status cache (this is because it also wants to keep track of things that were previous requested as "not missing", something the new system handles in a more systematic way in this very stack).

When you update the partitions definition, this makes the AssetSubset get deserialized as an empty subset instead of None (again, this only happens in the legacy system), meaning the system now things there are zero partitions which are not missing!

However, the specific test happened to be managing to succeed despite this bug because the asset under test had no dependencies. So its cursor was always set to the beginning of time, and so it'd start off with this empty set of things that had been handled already, but then add in all of the partitions which had been materialized since the cursor (which was locked to "the beginning of time"), meaning it would actually accidentally arrive at the right answer, and my fix to make the cursor advance correctly broke the test.

So I fixed the secondary bug as well by just having special handling when a partitions definition changes (and to emphasize, this "treat an invalid partitions subset as an empty partitions subset" is terrible because of the possibility of these bugs and has been completely divested from in the AssetSlice world)

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21973?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#21974** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21974?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21973** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21973?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#21972** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21972?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21971** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21971?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21853** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21853?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#21854](https://github.com/dagster-io/dagster/pull/21854) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21854?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#21844** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21844?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21839** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21839?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/declarative_scheduling/scheduling_condition.py [hunk 1]
```diff
@@ -185,6 +186,13 @@ def parent_newer() -> "ParentNewerCondition":
 
         return ParentNewerCondition()
 
+    @staticmethod
+    def newly_updated() -> "NewlyUpdatedCondition":
+        """Returns a SchedulingCondition that is true for an asset partition if it has been updated since the previous tick."""
+        from .operands import NewlyUpdatedCondition
+
+        return NewlyUpdatedCondition()
+
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/declarative_scheduling/scheduling_condition.py:195]: Hmm given the description here I am now confused how this relates to the more general "SinceCondition" down the stack...

**Review by schrockn — CHANGES_REQUESTED** [review]
q mgmt

File: python_modules/dagster/dagster/_core/definitions/declarative_scheduling/operands/slice_conditions.py [hunk 2]
```diff
@@ -99,6 +99,22 @@ def compute_slice(self, context: SchedulingContext) -> AssetSlice:
         )
 
 
+@whitelist_for_serdes
+class NewlyUpdatedCondition(SliceSchedulingCondition):
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/declarative_scheduling/operands/slice_conditions.py:103]: `UpdatedOnePreviousTick` or something? I think it is important that we understand that this is only true one tick/tick interval at a time.

**Review by schrockn — APPROVED** [review]

**Comment by OwenKephart** [comment]
### Merge activity

* **Jun 4, 9:49 AM EDT**: @OwenKephart started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/21973).

**Pull request merged**